tuf/repository_lib.py

```diff
@@ -1284,43 +1304,26 @@ def generate_targets_metadata(targets_directory, target_files, version,
 
   # Ensure the user is aware of a non-existent 'target_directory', and convert
   # it to its abosolute path, if it exists.
-  targets_directory = _check_directory(targets_directory)
-
-  # Generate the fileinfo of all the target files listed in 'target_files'.
-  for target, custom in six.iteritems(target_files):
+  check_exists = not use_existing_fileinfo
+  targets_directory = _check_directory(targets_directory, check_exists)
 
-    # The root-most folder of the targets directory should not be included in
-    # target paths listed in targets metadata.
-    # (e.g., 'targets/more_targets/somefile.txt' -> 'more_targets/somefile.txt')
-    relative_targetpath = target
+  if use_existing_fileinfo:
+    for target, fileinfo in six.iteritems(target_files):
 
-    # Note: join() discards 'targets_directory' if 'target' contains a leading
-    # path separator (i.e., is treated as an absolute path).
-    target_path = os.path.join(targets_directory, target.lstrip(os.sep))
-
-    # Ensure all target files listed in 'target_files' exist.  If just one of
-    # these files does not exist, raise an exception.
-    if not os.path.exists(target_path):
-      raise securesystemslib.exceptions.Error(repr(target_path) + ' cannot'
-        ' be read.  Unable to generate targets metadata.')
+      # Ensure all fileinfo entries in target_files have a non-empty hashes dict
+      if not fileinfo.get('hashes', None):
+        raise securesystemslib.exceptions.Error('use_existing_hashes option set'
+            ' but no hashes exist in roledb for ' + repr(target))
 
-    # Add 'custom' if it has been provided.  Custom data about the target is
-    # optional and will only be included in metadata (i.e., a 'custom' field in
-    # the target's fileinfo dictionary) if specified here.
-    custom_data = None
-    if len(custom):
-      custom_data = custom
+      if fileinfo.get('length', -1) < 0:
```

Comment:
Cool way of checking this. 👍

Reply:
Feels a bit C-like, but I couldn't think of a cleaner way to do it.